### DISCUSSION TITLE
How to apply setter injection through attribute?

### DISCUSSION THREAD

**Discussion by fd6130** [symfony/symfony#45488] [Q&A] [answered]
https://symfony.com/doc/current/service_container/calls.html

Hi, I have follow the docs from above using attribute instead of phpDoc and my property showing null, no service being injected through setter method. I'm using Symfony 5.3 .

```php
trait Foo 
{
   private $container;

   #[Required]
   public function setContainer(ContainerInterface $container)
   {
       $this->container = $container;
   }

   public function getContainer()
   {
       return $this->container;
   }
}


class Bar
{
   use Foo;
}


// using phpDoc
dd($bar->getContainer()); // return ContainerInterface


// using attribute
dd($bar->getContainer()); // return null
```

**Comment by ahundiak** [comment, marked as answer]
How are you instantiating $bar?  It's not uncommon to see code such as `$bar = new Bar();`. Which of course will not work. 

You need to either inject bar or pull it from the container.  You also need to verify that Bar itself it properly defined as a service.

**Reply by fd6130** [reply]
I inject $bar through constructor in my controller.

```php
// In controller class
public function index(Bar $bar)
{
  dd($bar->getContainer()); // return null if using attribute
}
```

**Reply by ahundiak** [reply]
Okay.  So it's probably a question of how Bar was wired.  Does `bin/console debug:container Bar` show a call to setContainer?  If not then you probably need to enable autowire.

**Reply by ahundiak** [reply]
Actually you are probably just missing a use statement for the attribute:
```
# Foo.php
namespace App;

use Psr\Container\ContainerInterface;
use Symfony\Contracts\Service\Attribute\Required;

trait Foo 
{
   private $container;

   #[Required]
   public function setContainer(ContainerInterface $container)
   {
       $this->container = $container;
   }

   public function getContainer()
   {
       return $this->container;
   }
}
```
My VSCode IDE indicated something was wrong with the Required attribute but did not offer to insert a use statement to fix it.  Got that from the docs.

Furthermore you need to explicitly define an alias for the service container:
```
# services.yaml
services:
...
    Psr\Container\ContainerInterface $container: '@service_container'
```
Injecting the complete container is of course discouraged.  Hence the need to explicitly define the alias.

**Reply by fd6130** [reply]
Thanks. It is my code issue. And your solution is works for me when using attribute as well.
